backup_control: Fix backup mode determination

Funky185540 opened the pull request:
Fixes determination of the backup mode, as specified in the [GNU manual][1].

Previously the backup mode determination wouldn't allow abbreviations to the backup mode, and didn't allow to differ between the short `-b` option (which automatically defaults to the `"existing"` backup style) and the long `--backup` option (which has more special handling).

Also, handling of the `--backup` option has been revamped such that it can now be passed without an argument to it (in which case the `VERSION_CONTROL` env variable is read, and if that doesn't exist it defaults to `"existing"`). To achieve this the separator between this option and a following argument, if present, must be an equals-sign ("="). This is required in order to supply `--backup` without arguments (i.e. as a flag).

Due to this, previously it was impossible to use the option as a flag, as it expected arguments to be space-delimited. I.e. a syntax like `... --backup foo bar`, where *foo* and *bar* are meant as arguments to the utility (i.e. file arguments to mv) would have consumed *foo* as backup style, which isn't meant by the user nor valid.

Also modifies the backup-specific help-texts and adds error outputs for invalid or ambiguous backup modes.


[1]: https://www.gnu.org/software/coreutils/manual/html_node/Backup-options.html

Funky185540 commented on src/uucore/src/lib/mods/backup_control.rs:
```diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
+        } else {
+            "ambiguous"
+        };
+        Err(format!(
```
Also I'm not quite happy with the formatting of this bit in the code, as it ruins the indentation. I'd be thankful for suggestions on how to improve it!

miDeb replied:
I think it's fine. The only thing I could think of is to move the "Valid arguments..." part to a constant, but I don't think that's necessary or useful.

miDeb commented on src/uucore/src/lib/mods/backup_control.rs:
````diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
+/// };
+/// ```
+pub fn determine_backup_mode(
+    sopt_present: bool,
+    lopt_present: bool,
+    lopt_value: Option<&str>,
````
nit: could you use a longer name for `sopt` and `lopt`, like `long_opt` and `short_opt`? That's a bit more obvious when reading the code.

Funky185540 replied:
Done!

miDeb commented on src/uucore/src/lib/mods/backup_control.rs:
```diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
+            "numbered" | "t" => Ok(BackupMode::NumberedBackup),
+            "existing" | "nil" => Ok(BackupMode::ExistingBackup),
+            "none" | "off" => Ok(BackupMode::NoBackup),
+            _ => panic!(), // cannot happen as we must have exactly one match
```
nit: the `unreachable!()` macro might be a better fit here because it makes it clear that this can never be reached.

Funky185540 replied:
Cool, didn't know that exists! Added it.

miDeb commented on src/uucore/src/lib/mods/backup_control.rs:
```diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
+/// invalid, an error is returned. The error contains the formatted error string
+/// which may then be passed to the [`show_usage_error`] macro.
+fn match_method(method: &str, origin: &str) -> Result<BackupMode, String> {
+    let x = vec![
```
This doesn't need to be a `Vec` but could be a plain array. (i.e. you can remove `vec!` and everything should still work.)

Funky185540 replied:
Using the definitions from `BACKUP_CONTROL_VALUES` instead now, which already exists in this module.

miDeb commented on src/uucore/src/lib/mods/backup_control.rs:
```diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
         }
     } else {
-        BackupMode::NoBackup
+        let error_type = if matches.len() == 0 {
```
nit: you can use `matches.is_empty()` here.

miDeb reviewed (commented):
From my perspective this looks good!
I just have some minor suggestions.
It would be great if you could add some tests as well to make sure everything works (and keeps working) correctly :)

Funky185540 commented:
> From my perspective this looks good!
> I just have some minor suggestions.
> It would be great if you could add some tests as well to make sure everything works (and keeps working) correctly :)

Thanks for the feedback!

I tried to incorporate the tests into a `tests` module inside `backup_control` (Because I thought that it just fits nicely there). I have marked it with `#[cfg(test)]` and marked all test functions with `#[test]` but for some reason the test functions don't show up in the output of cargo test. Is there something else I need to consider to add the test, do I need to explicitly "register" them somewhere?

miDeb commented:
The tests are not run because this repo is structured as a workspace of multiple crates. If you run `cargo test` only the tests in the top level crate (`tests/*`) will be run. To run all tests in the workspace you can run `cargo test --workspace`. I think the easier way would be to `cd src/uucore` first and run `cargo test` from there. This will only run the tests in `uucore`.

Funky185540 commented:
Thanks a lot!

2 Issues remain with the tests:

1. My doctest doesn't find clap, and I can't tell why
2. Some of the tests use env-variables, and it seems to me that these are shared between the test runner threads. Is there a better way to declare environment variables in tests? Or is there a directive that tells rust to run the tests in this module in series rather than parallel?

miDeb commented:
> My doctest doesn't find clap, and I can't tell why

I think that's because clap isn't a dependency of `uucore`. You can add it as a dev-dependency in `Cargo.toml` (the one in uucore).

>  Is there a better way to declare environment variables in tests? Or is there a directive that tells rust to run the tests in this module in series rather than parallel?

I don't know of any of that. A way to test this would be to extract the logic of determining the backup mode into a private function that takes the environment variable as an argument. You can then test that function.

To test with an actual environment variable you could add integration tests for the utilities that use `backup_control`. Integration tests live in `tests/by-util/<utility>`. The `ucmd` thing that you can see there frequently creates a new process and you can do `ucmd.env()` to set an environment variable just for that process.

Funky185540 commented:
I think I fixed it now. I did some research on this type of issue and found this discussion: https://github.com/rust-lang/rust/issues/47506
One of the solutions proposed in there is the use of [the rusty-fork crate][1], which is what I have gone for now. Alternatively I could try to use `lazy_static` to setup a shared mutex in the test submodule and lock the mutex before running the tests.

I mostly went for rusty-fork as I wasn't aware when implementing the Mutex that one can import single macros in Rust 2018 -- because rust would refuse to compile my code when I included `#[macro_use] extern crate lazy_static;`, as I didn't import the macros at the crate root.

If you'd like me to change this I can try that. I'm actually not sure about the cross-platform support that rusty-fork offers...


[1]: https://docs.rs/rusty-fork/0.3.0/rusty_fork/

Funky185540 commented:
I have now got a solution that works with a global Mutex (scoped for the test submodule only) instead. Both work fine it seems, so i'll leave the final choice up to you. It *feels* to me like the Mutex-solution is more cross-platform compatible, but that's just a guess.

Funky185540 commented:
Rebased onto master. I have noticed that the `test_du_bytes` test fails, but that's unrelated to my changes as it happens on the master branch as well.

I have now implemented the Mutex-based approach as it's more cross-platform compatible. This requires me to lock the mutex in a few tests, so I went ahead and lock it in all tests just to make sure I don't miss one. As a consequence these tests are now run in series instead of parallel, but due to their low cost it's hardly noticable.

I have also adapted the `install` utility as it got support for `--backup` & co. with pull #2457 .

Funky185540 commented:
Build/Makefile failed a test in cp, this isn't related to this PR.

Funky185540 commented:
@miDeb Do you have some more thoughts on the current implementation?

miDeb commented:
Sorry, I didn't see your last comment. I will take a look now

miDeb commented on src/uu/cp/src/cp.rs:
```diff
@@ -604,9 +606,19 @@ impl Options {
+        let backup_mode = match backup_mode {
+            Err(err) => {
+                return Err(Error::Backup(err));
+                // show_usage_error!("{}", err);
+                // return 1;
```
Do you still need this comment?

Funky185540 replied:
Nope.

miDeb commented on src/uu/install/src/install.rs:
```diff
@@ -308,15 +308,25 @@ fn behavior(matches: &ArgMatches) -> Result<Behavior, i32> {
         None
     };
 
+    let backup_mode = backup_control::determine_backup_mode(
+        matches.is_present(OPT_BACKUP_NO_ARG),
+        matches.is_present(OPT_BACKUP),
+        matches.value_of(OPT_BACKUP),
+    );
```
What would you think about making `determine_backup_mode` take the whole`matches` instead of three arguments? (I think this version is fine though, too)

Funky185540 replied:
I already thought about it and that'd be cool, but in that case I'd have to streamline the naming of the various arguments involved across the utilities... I'll try to find the time tomorrow.

Example: 

- install calls the arguments: "backup", "backup2", "suffix"
- cp calls the arguments: "backup", "b", "suffix"

Funky185540 replied:
You know I gave this some more thought. I'd go even one step further with this: All utils that implement `backup`s functionality have *exactly the same options for that*. So I'd propose that instead of defining the options in each of the applications separately, I may just as well define the backup-related options in the `backup_control` as "finished" clap-Arguments (with all the texts and flags etc.) and then add these to the App in each of the applications. Then I can pass in all `matches` to `determine_backup_mode`, throw errors if the flags weren't present and continue as before if all is well and defined properly.

I think that I'd like to that in a separate PR though, as this is already somewhat big and I'm afraid I may lose myself in the changes... Would that be okay for you or should I attempt to do it in this PR?

miDeb replied:
Agree, doing this in a new PR is fine

miDeb commented on src/uu/cp/src/cp.rs:
```diff
@@ -604,9 +606,19 @@ impl Options {
+        let backup_mode = match backup_mode {
+            Err(err) => {
+                return Err(Error::Backup(err));
+                // show_usage_error!("{}", err);
```
That's not something you have to address now, but it would be interesting to integrate this with our new error handling mechanism at a later point when all utilities support it, instead of having to repeat the error handling in all utilities.

Funky185540 replied:
Yes, definitely! I think I'll add it as a `UCommonError`, as it is consistent across utilities.

miDeb reviewed (approved):
LGTM, sorry for the delay. Thanks for working on this!

Funky185540 commented:
Okay, I'll start working on a new PR to:

- Streamline the backup-control related argument handling
- Add a UCustomError for Backup-related errors

Unless you have other suggestions to this PR it can be merged now. :)
